python/whylogs/experimental/core/metrics/embedding_metric.py

```diff
@@ -0,0 +1,204 @@
+
+        if (
+            self.labels != other.labels
+            or self.distance_fn != other.distance_fn
+            or not (self.references.value == other.references.value).all()
+        ):
+            raise ValueError("Attempt to merge incompatible EbeddingMetrics")
+
+        result = EmbeddingMetric(self.references, self.labels, self.distance_fn, self.serialize_references)
+        result.submetrics = self.merge_submetrics(other)
+        return result
+
+    def to_protobuf(self) -> MetricMessage:
+        msg = {}
+        for sub_name, metrics in self.submetrics.items():
+            for namespace, metric in metrics.items():
+                sub_msg = metric.to_protobuf()
+                for comp_name, comp_msg in sub_msg.metric_components.items():
+                    msg[f"{sub_name}:{namespace}/{comp_name}"] = comp_msg
+        if self.serialize_references:
+            msg["references"] = self.references.to_protobuf()
+
+        return MetricMessage(metric_components=msg)
+
+    def _update_submetrics(self, submetric: str, data: PreprocessedColumn) -> None:
+        for key in self.submetrics[submetric].keys():
+            self.submetrics[submetric][key].columnar_update(data)
+
+    def columnar_update(self, data: PreprocessedColumn) -> OperationResult:
+        if data.numpy.ints is None and data.numpy.floats is None:
+            return OperationResult.ok(0)
+
+        matrices = [X for X in [data.numpy.ints, data.numpy.floats] if X is not None]
+        for X in matrices:  # TODO: throw if not 2D ndarray
+            X_ref_dists = self.distance_fn(X, self.references.value)  # type: ignore
+            X_ref_closest = np.argmin(X_ref_dists, axis=1)
+            print(f"ref dists: {X_ref_dists}")
+            print(f"closest: {X_ref_closest}")
+            for i in range(X_ref_dists.shape[1]):
+                print(f"updating {self.labels[i]}_distance with {X_ref_dists[ :, i].tolist()}")
+                self._update_submetrics(
+                    f"{self.labels[i]}_distance", PreprocessedColumn.apply(X_ref_dists[:, i].tolist())
+                )
+
+            closest = [self.labels[i] for i in X_ref_closest]
+            print(f"updating closest with {closest}")
+            self._update_submetrics("closest", PreprocessedColumn.apply(closest))
+
+        return OperationResult.ok(len(matrices))
+
+    @classmethod
+    def from_protobuf(cls, msg: MetricMessage) -> "EmbeddingMetric":
+        if "references" in msg.metric_components:
+            references = MatrixComponent.from_protobuf(msg.metric_components["references"])
+            msg.metric_components.pop("references")
+            serialize_references = True
+        else:
+            references = np.zeros((1, 1))
+            serialize_references = False
+
+        submetrics = EmbeddingMetric.submetrics_from_protobuf(msg)
+
+        # TODO: this doesn't guarantee order :(  Fix it, or don't [de]serialize
```

Comment:
Do we need to deserialize the labels in reference vector order, i.e., maintain the relationship between labels and reference matrix rows? If so, we'll probably need to serialize them explicitly rather than recovering them from the submetric names. Deserialized metrics aren't updatable, but they are supposed to be mergeable. I'd have to remove the? same-label check in the merge code if we can't guarantee the label ordering. Do we need/want the labels for visualization

Reply:
We could sort the labels and then apply the same permutation to the reference matrix rows

Reply:
We will need to maintain relationship for visualization and interpretation purposes. Sorted labels should be fine. This seems similar to what happened in model performance metric serialization / merging, so you may check there.